### PR TITLE
Improve failsafe response to loss of position and velocity accuracy

### PR DESCRIPTION
**Pull request by priseborough**
Addresses a significant vulnerability to bad compass alignment or calibration that can cause a high speed flyaway on copters.

Replaces #6656 

Submitted for initial review and feedback.

TESTING

SITL testing performed with fallback from POSCTL and auto takeoff.
Loss of position estimation after takeoff in testing was induced by a 90 degree magnetic declination error added to the ecl EKF code.

### CODE REVIEW

**Comment by priseborough** [comment]
I ran into some issues with the rebase. This PR will be re-opened when they are fixed.

**Comment by LorenzMeier** [comment]
@priseborough Need further integration support or are you good now?

**Comment by priseborough** [comment]
The missing patch has been added. I have tested this in the copter jmavsim SITL verified entry and exit from failsafe in the following modes that require position and velocity:

Position
Hold
Mission
Return
Follow Me

The testing used the following patch to force the ekf2 module to report bad position and velocity accuracy:

https://github.com/priseborough/Firmware/commit/f627cc6a0a942b2e02273a2b57a39dc1306a3fbf

A test of the ability to protect against bad compass alignment was repeated by setting the following parameters in the EKF:

EKF2_DECL_TYPE = 4
EKF2_MAG_DECL = 90

This showed that there is still a race condition with the ekf reset logic and the logic in the commander that can occur which needs to be investigated and fixed before this is merged.

**Comment by priseborough** [comment]
The race condition was traced to a combination of the bad position failsafe not being 'sticky' enough and the optimistic EKF reporting position accuracy during dead-reckoning. The commnader has been modified to make the invalid position failsafe more persistent and the EKF position accuracy reporting has been modified with this pending PR: https://github.com/PX4/ecl/pull/252

**Comment by LorenzMeier** [comment]
PR merged and you can update the submodule to fix this.

**Comment by LorenzMeier** [comment]
@priseborough SITL test still failing unfortunately.

**Comment by priseborough** [comment]
@LorenzMeier  I worked through the SITL log to try and determine why the test failed. It takes off OK and stays in off-board mode at a constant 2m altitude for the duration of the log. There is a short period just prior to arming where the failsafe takes it back to descend mode but I have been unable to determine why it stays in offboard after takeoff without responding to commands. At this point I am blocked because I do not know how to replicate the SITL test in my build environment.

**Comment by priseborough** [comment]
@LorenzMeier After running the tests documented here https://dev.px4.io/advanced-jenkins-ci.html#test-execution  on my dev environment using both this branch rebased on master and master at a5fa4e9, both are failing to complete CI SITL tests but with a different failure to the build server test http://sitl01.dronetest.io/job/PX4%20Firmware%20POSIX%20SITL%20PRs/4862/ .

[px4.rosunit-mission_test_old_2/test_mission][FAILURE]--------------------------
(vtol_old_2.txt) landing not detected after landing WP timeout: 600, index: 5
  File "/usr/lib/python2.7/unittest/case.py", line 331, in run
    testMethod()
  File "/job/catkin/src/px4/integrationtests/python_src/px4_it/mavros/mission_test.py", line 365, in test_mission
    self.wait_on_landing(600, index)
  File "/job/catkin/src/px4/integrationtests/python_src/px4_it/mavros/mission_test.py", line 265, in wait_on_landing
    (self.mission_name, timeout, index))
  File "/usr/lib/python2.7/unittest/case.py", line 424, in assertTrue
    raise self.failureException(msg)

Inspection of the ulog shows that the ecl ekf2 estimator velocity and position keep drifting away from the GPS and are having to be reset. I have never used the ROS simulation before and am unfamiliar with the sensor interface so at this point in time have not been able to determine the root cause.

In summary I have been unable to replicate the build server failure, but instead have a different test failure that remains unexplained.

**Comment by priseborough** [comment]
I tracked down the cause of the CI SITL test error - the script was attempting to arm the vehicle at 3.5 seconds from boot which was not long enough for the position accuracy checks to pass (they require 5 seconds of good data). I added a 5 second pause before arming and the checks now pass.

**Comment by priseborough** [comment]
After further testing I think I need to add a method to latch failsafe on until landed entered from an automatic takeoff mode or within TBD seconds of takeoff. In some tests the failsafe clears just before touchdown which results in a crash.

**Comment by priseborough** [comment]
After still more testing the following design requirements have been identified:

1) The failsafe logic that responds to an invalid position and velocity estimate should be slow to revert to the previous state if the error is detected during or shortly after takeoff because the errors are likely caused by an invalid heading estimate that will cause repeated GPs rejection and resets.
2) If the pilot or ground station issues a mode change command requiring GPS and the EKF innovation checks are passing again, the probation period should be cancelled and the mode change actioned.
3) An algorithm that increases the probation period during test failure and decreases it slowly during test pass should be used to ensure that an error that causes repeat failures will end up with the nav being declared unhealthy for long enough for the vehicle to complete a failsafe landing/descent.

I have a prototype for items 1) and 3). Item 2) requires more investigation.

**Comment by priseborough** [comment]
I have rebased and made the following modifications:

- Local and global position validity tests have been refactored and moved into separate functions.
- The concept of a variable probation period has been introduced. Probation is the time required for tests to pass before an invalid position or velocity status can be declared valid. 
- When on-ground the probation period is set to a small fixed value. 
- At takeoff the probation period  is set to a larger value. This allows time for the vehicle to complete a failsafe landing if taking off with bad sensors which causes the EKF intermittently declares itself good when resetting to GPS. 
- When in-flight, the probation period increases at a multiple of real time when tests are failing and decreases at real time when tests are passing. This means that short duration errors such as an isolated GPS glitch during flight will recover faster.
- A reset function has been added which resets the probation interval and re-runs the tests. This reset is performed whenever the user demands a mode change. This enables the user to regain control faster following a loss of estimation.

5.

**Comment by priseborough** [comment]
I don't think this should wait for the commander re-write because it addresses a significant vulnerability to misaligned in badly calibrated magnetometers that can cause flyaways.

**Comment by priseborough** [comment]
@LorenzMeier This is ready for final review and wider testing. I have flight tested it by rotating the magnetometer by 90 degrees to create an extreme yaw error, and then performed a takeoff in position  control mode. It auto reverts to alt control. All testing in auto modes been done in SITL.

**Comment by priseborough** [comment]
This has been forked by https://github.com/PX4/Firmware/pull/7095 which is currently under review

**Comment by santiago3dr** [comment]
few flights on V3 (pixhawk mini); good position control; no issues
http://logs.px4.io/plot_app?log=3b43cb5b-8a2a-473c-8bca-9d405624c49c
http://logs.px4.io/plot_app?log=c5a2215f-76b7-4614-b4bf-8d5b7d3f86c2
http://logs.px4.io/plot_app?log=0bea76fc-973f-4c54-b5ef-b299c47dd107
